### PR TITLE
Added pagination to the table

### PR DESCRIPTION
**Pull request by SolomiiaLomnytska22**
Refs:
#21 (Add pagination to table data)

Proposed changes:
- Added pagination on the bottom of the page
- Created customisable pagination component
- Added ability to select the quantity of rows to display
- Pagination component contains three page buttons to show current page, next and previous, navigation buttons "<" or ">" and buttons that navigate to the first and last pages.

User interface before implementing changes:

Screenshot 2024-01-25 at 12 08 29

User interface after implementing changes:

Screenshot 2024-01-25 at 12 09 06
Screenshot 2024-01-25 at 12 17 54

Verification:
- Open home page with table
- Load data from json server or add new entries using "+ Add" button
- Scroll to the bottom of the page
- Enter the quantity of table rows to display
- Press on the button with page number to navigate to this page
- Press Screenshot 2024-01-25 at 12 16 00 button to navigate to the previous page
- Press Screenshot 2024-01-25 at 12 16 40 button to navigate to the next page

All changes tested locally. New and already existing functionality is supported.

https://github.com/SolomiiaLomnytska22/solomiia-training/assets/136481621/62e7e1d7-1fb9-456c-baec-d92d45e1095d

### CODE REVIEW

File: app/src/components/common/Button.vue [hunk 1]
```diff
@@ -65,4 +65,17 @@ button {
...
+  border-color: #ccc;
+  color: #666;
+  cursor: not-allowed;
+}
+
+.disabled:hover {
+  background: #ccc;
+  border-color: #ccc;
+  color: #666;
+}
```
Comment (kaytee-te) [inline app/src/components/common/Button.vue]: Maybe you find it useful.

```suggestion
.disabled,
.disabled:hover {
  background: #ccc;
  border-color: #ccc;
  color: #666;
}

/* Specific property for the normal (non-hover) state */
.disabled {
  cursor: not-allowed;
}
```

File: app/src/components/common/Pagination.vue [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+      label="Rows per page"
+      :min="1"
+      :max="totalRows"
+      :value="rowsPerPage"
+      type="number"
+      @input="updateRowsPerPage"
+    />
+    <div class="pages">
+      <Button
+        :style-type="currentPage === 1 ? 'disabled' : 'secondary'"
```
Comment (kaytee-te) [inline app/src/components/common/Pagination.vue]: Let's move this condition into getter.

**Comment by kaytee-te** [comment]
Your function names are intuitive and make it easy to understand the purpose of each function. Great job!

File: app/src/components/common/Button.vue (hunk 1 above)
Comment (marina-grechukha) [reply app/src/components/common/Button.vue]: I dont think the style of disabled button on hover should be different

File: app/src/components/common/Button.vue [hunk 3]
```diff
@@ -18,7 +18,7 @@ export default class Button extends Vue {
   @Prop({
     required: false,
     default: 'primary',
-    validator: (value: string) => [ 'primary', 'secondary' ].includes(value)
+    validator: (value: string) => [ 'primary', 'secondary', 'disabled' ].includes(value)
```
Comment (marina-grechukha) [inline app/src/components/common/Button.vue]: 'disabled'  it's not really a style type, I think it's better to make it a separate prop

File: app/src/components/common/__tests__/NumberInput.spec.ts [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+  it('renders label and input with correct props', () => {
+    expect(wrapper.find('label').text()).toBe(`${wrapper.props('label')}:`)
+    expect(wrapper.find('input').attributes('id')).toBe(wrapper.props('id'))
+    expect((wrapper.find('input').element as HTMLInputElement).value).toBe(wrapper.props('value').toString())
+  })
+
+  it('emits input event when input value changes', async () => {
+    const inputField = wrapper.find('input')
+    ;(inputField.element as HTMLInputElement).value = '30'
```
Comment (marina-grechukha) [inline app/src/components/common/__tests__/NumberInput.spec.ts]:
```suggestion
    (inputField.element as HTMLInputElement).value = '30'
```

File: app/src/components/common/Button.vue (hunk 1 above)
Comment (SolomiiaLomnytska22) [reply app/src/components/common/Button.vue]: Fixed repeating style classes for the button in the recent commit.

File: app/src/components/common/Pagination.vue (hunk 2 above)
Comment (SolomiiaLomnytska22) [reply app/src/components/common/Pagination.vue]: Moved condition to the getters in the recent update

File: app/src/components/common/Button.vue (hunk 3 above)
Comment (SolomiiaLomnytska22) [reply app/src/components/common/Button.vue]: Thanks for advice! Made separate prop for it in the recent commit.

File: app/src/components/common/__tests__/NumberInput.spec.ts (hunk 4 above)
Comment (SolomiiaLomnytska22) [reply app/src/components/common/__tests__/NumberInput.spec.ts]: Hi! I get an error message from the ESLint: `Unexpected newline between function and ( of function call.(no-unexpected-multiline)` when this semicolon is deleted. It seems that I have encountered a bug similar to this one: https://github.com/eslint/eslint/issues/16183 and I don't really know how to fix this issue.

Comment (SolomiiaLomnytska22) [reply app/src/components/common/__tests__/NumberInput.spec.ts]: Hey, @kaytee-te and @marina-grechukha! I've tried your solution for the issue in NumberInput.spec.ts 
(`wrapper.find('input').trigger('input', {  value: '30' })`), but you were right and in such case my input does not emit the correct value. However, I fixed it by using setValue function which is equivalent to those two problematic lines of code. Here is a snippet from the documentation.
Screenshot 2024-01-31 at 10 00 43
Hope that this solution is fine)

**Comment by SolomiiaLomnytska22** [comment]
Hi, @marina-grechukha and @kaytee-te, in the recent commit I moved pagination setup from the Users.vue page to my Table.vue component!

File: app/src/Users.vue [hunk 5]
```diff
@@ -73,10 +73,12 @@ import ConfirmationDialog from '@/components/common/ConfirmationDialog.vue'
 import Toast from './components/common/Toast.vue'
 import { Ref } from 'vue-property-decorator'
 import TextInput from '@/components/common/TextInput.vue'
+import Pagination from '@/components/common/Pagination.vue'
```
Comment (marina-grechukha) [inline app/src/Users.vue]: I think you don’t need this here anymore

File: app/src/Users.vue [hunk 6]
```diff
@@ -73,10 +73,12 @@ import ConfirmationDialog from '@/components/common/ConfirmationDialog.vue'
...
 
 @Component({
   components: {
     TextInput,
+    Pagination,
```
Comment (marina-grechukha) [inline app/src/Users.vue]: and this ;)

File: app/src/Users.vue (hunk 5 above)
Comment (SolomiiaLomnytska22) [reply app/src/Users.vue]: Removed this import in the recent commit!

File: app/src/Users.vue (hunk 6 above)
Comment (SolomiiaLomnytska22) [reply app/src/Users.vue]: Fixed it in the recent commit

**Review by marina-grechukha — APPROVED** [review]
Looks good 👍

**Review by kaytee-te — APPROVED** [review]

**Pull request merged**